documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_moderation_metrics.md

```diff
@@ -0,0 +1,334 @@
+- report rejected
+- report deduplicated
+- decision created
+
+```{todo}
+Find answers to these questions:
+
+- Why does the
+  [Django Admin IP](/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.md)
+  have separate events for `confirmed_sensitive` and `deindexed_sensitive`?
+
+- Additionally shouldn't the action just be `confirmed` (with deindexing being
+  a consequence of that action)?
+```
+
+The log line must contain a JSON that matches this type definition.
+
+```typescript
+interface ReportMessage {
+  message_type: "ModerationReport"
+  event_type: "created" | "confirmed" | "rejected" | "deduplicated"
+  media_type: "image" | "audio"
+}
+```
+
+```typescript
+interface DecisionMessage {
+  message_type: "ModerationDecision"
+  event_type: "created"
+  action_type: "confirmed" | "rejected" | "deduplicated"
+  media_type: "image" | "audio"
+}
+```
+
+This logging can be abstracted into a specific utility module in the API. These
+utilities must be invoked from the `save` method on the Report and Decision
+models. This will ensure that these events are logged regardless of whether
+moderation happens via a view (like the admin site) or via a console interface
+(like a management command).
+
+### Surface in AWS CloudWatch
+
+Real-time metrics will be delivered to CloudWatch via Logs Insights. We can then
+create a dashboard in CloudWatch to visualise these metrics.
+
+Here is a sample Logs Insights query that represents the volume of reports and
+our various responses.
+
+```text
+fields @timestamp, message_type, event_type
+| filter message_type like "ModerationReport"
+| stats
+    sum(case when event_type = 'created' then 1 else 0 end) as createdCount,
+    sum(case when event_type != 'created' then 1 else 0 end) as moderatedCount,
+    sum(case when event_type = 'confirmed' then 1 else 0 end) as confirmedCount,
+    sum(case when event_type = 'rejected' then 1 else 0 end) as rejectedCount,
+    sum(case when event_type = 'deduplicated' then 1 else 0 end) as duplicateCount,
+  by bin(5m)
+| sort @timestamp desc
+```
+
+```text
+fields @timestamp, message_type, event_type
+| filter message_type like "ModerationDecision"
+| stats
+    sum(case when event_type = 'created' then 1 else 0 end) as createdCount,
+    sum(case when action_type = 'confirmed' then 1 else 0 end) as confirmedCount,
+    sum(case when action_type = 'rejected' then 1 else 0 end) as rejectedCount,
+    sum(case when action_type = 'deduplicated' then 1 else 0 end) as duplicateCount,
+  by bin(5m)
+| sort @timestamp desc
+```
+
+This query can be used to create a line chart that shows the volume of reports
+and our various responses over time. Similar queries can be used to create
+charts for other metrics. Ultimately a dashboard can be put together in
+CloudWatch, that tracks all relevant metrics.
+
+```{todo}
+Should we have this dashboard tracked in the infrastructure repository?
+```
```

Comment:
Addressed to the @WordPress/openverse-infrastructure team.

Reply:
Any new dashboard should be tracked once we've finalised its format, if for no other reason than so that we can easily recover it in case the live dashboard gets deleted.